mixer/cmd/versions.go

```diff
@@ -0,0 +1,141 @@
+
+Without any flags, it increments mix by 10 and upstream to the latest
+version of the same format used by current upstream. Use --mix-version
+and --upstream-version to set a specific version. By default the
+versions are shown but not set, to set use --set (-s).
+
+This command is aware of format bumps and will not automatically
+propose a version change cross different formats in upstream. At the
+moment format bumps must be handled manually when using Mixer.
+`,
+	Run: runUpdateVersions,
+}
+
+var updateVersionsFlags struct {
+	mixVersion      uint32
+	upstreamVersion uint32
+	set             bool
+}
+
+func init() {
+	RootCmd.AddCommand(updateVersionsCmd)
+	updateVersionsCmd.Flags().StringVarP(&config, "config", "c", "", "Builder config to use")
+	updateVersionsCmd.Flags().Uint32Var(&updateVersionsFlags.mixVersion, "mix-version", 0, "Next mix version (0 means increment current one)")
+	updateVersionsCmd.Flags().Uint32Var(&updateVersionsFlags.upstreamVersion, "upstream-version", 0, "Next upstream version (0 means take the latest possible)")
+	updateVersionsCmd.Flags().Uint32Var(&updateVersionsFlags.upstreamVersion, "clear-version", 0, "Alias for --upstream-version")
+	updateVersionsCmd.Flags().BoolVarP(&updateVersionsFlags.set, "set", "s", false, "Set the new versions in the mix")
+}
+
+func runUpdateVersions(cmd *cobra.Command, args []string) {
+	b, err := builder.NewFromConfig(config)
+	if err != nil {
+		fail(err)
+	}
+
+	nextMix := updateVersionsFlags.mixVersion
+	nextUpstream := updateVersionsFlags.upstreamVersion
+
+	currentMix, err := parseUint32(b.MixVer)
+	if err != nil {
+		failf("Couldn't parse current mix version %q: %s", b.MixVer, err)
+	}
+
+	currentUpstream, err := parseUint32(b.UpstreamVer)
+	if err != nil {
+		failf("Couldn't parse current upstream version %q: %s", b.UpstreamVer, err)
+	}
+
+	if nextMix == 0 {
+		nextMix = currentMix + 10
+	}
+	if nextMix < currentMix {
+		failf("Next mix version (%d) must be greater than current mix version (%d)", nextMix, currentMix)
+	}
+
+	upstreamFormat, err := b.DownloadFileFromUpstream(fmt.Sprintf("update/%d/format", currentUpstream))
+	if err != nil {
+		failf("Couldn't download information about upstream: %s", err)
+	}
+
+	latest, err := b.DownloadFileFromUpstream(fmt.Sprintf("update/version/format%s/latest", upstreamFormat))
+	if err != nil {
+		failf("Couldn't download information about upstream: %s", err)
+	}
+
+	latestUpstream, err := parseUint32(latest)
+	if err != nil {
+		failf("Couldn't parse latest upstream version: %s", err)
+	}
+
+	switch {
+	case nextUpstream == 0:
+		nextUpstream = latestUpstream
+	case nextUpstream < currentUpstream:
+		failf("Next upstream version (%d) must be at least the current upstream version (%d)", nextUpstream, currentUpstream)
+	case nextUpstream > latestUpstream:
+		failf("Can't update upstream to %d. The latest upstream version in format %s is %d.", nextUpstream, upstreamFormat, latestUpstream)
```

Comment:
For now this is exactly the right thing to do. In the future (post auto-format-bumps), this is when we'd instead say `"Can't update upstream to %d without performing a format bump. The latest upstream version in format %s is %d.\nWould you like to apply a format bump now? [y/N]:"`

Reply:
That's the spirit, or even give a flag that tells the user it can bump regardless.

Note that one problem raised in previous discussions was: we would need to make sure metadata we generate for the "last version of a format" is compatible to the swupd running there. I think we are in a better position now to guarantee that (c.f. Icarus comments about why is good to generate manifests via templates) via mixer itself. See also https://github.com/clearlinux/mixer-tools/issues/122.